memtable/write_buffer_manager.cc

```diff
@@ -115,52 +147,26 @@ void WriteBufferManager::FreeMemWithCache(size_t mem) {
   s.PermitUncheckedError();
 }
 
-void WriteBufferManager::BeginWriteStall(StallInterface* wbm_stall) {
+bool WriteBufferManager::MaybeBeginWriteStall(StallInterface* wbm_stall) {
   assert(wbm_stall != nullptr);
-  assert(allow_stall_);
-
-  // Allocate outside of the lock.
   std::list<StallInterface*> new_node = {wbm_stall};
-
-  {
-    std::unique_lock<std::mutex> lock(mu_);
-    // Verify if the stall conditions are stil active.
-    if (ShouldStall()) {
-      stall_active_.store(true, std::memory_order_relaxed);
-      queue_.splice(queue_.end(), std::move(new_node));
-    }
-  }
-
-  // If the node was not consumed, the stall has ended already and we can signal
-  // the caller.
-  if (!new_node.empty()) {
-    new_node.front()->Signal();
+  std::unique_lock<std::mutex> lock(wbm_mutex_);
+  if (ShouldStall()) {
+    wbm_stall->SetState(StallInterface::State::BLOCKED);
+    queue_.splice(queue_.end(), std::move(new_node));
+    return true;
+  } else {
+    return false;
   }
 }
 
-// Called when memory is freed in FreeMem or the buffer size has changed.
 void WriteBufferManager::MaybeEndWriteStall() {
-  // Cannot early-exit on !enabled() because SetBufferSize(0) needs to unblock
-  // the writers.
-  if (!allow_stall_) {
+  if (ShouldStall()) {
     return;
   }
 
-  if (IsStallThresholdExceeded()) {
-    return;  // Stall conditions have not resolved.
-  }
-
-  // Perform all deallocations outside of the lock.
```

Comment:
Note: we can't do this now as we need to hold the lock for calling MaybeEndWriteStall(). See perf test result for any perf regression concern.